packages/datadog-plugin-express/src/code_origin.js

```diff
@@ -0,0 +1,32 @@
+'use strict'
+
+const { entryTags } = require('../../datadog-code-origin')
+const Plugin = require('../../dd-trace/src/plugins/plugin')
+const web = require('../../dd-trace/src/plugins/util/web')
+
+const kCodeOriginForSpansTagsSym = Symbol('datadog.codeOriginForSpansTags')
+
+class ExpressCodeOriginForSpansPlugin extends Plugin {
+  static get id () {
+    return 'express'
+  }
+
+  constructor (...args) {
+    super(...args)
+
+    this.addSub('apm:express:middleware:enter', ({ req, layer }) => {
+      const tags = layer[kCodeOriginForSpansTagsSym]
+      if (!tags) return
+      const context = web.getContext(req)
+      context.span?.addTags(tags)
+    })
+
+    this.addSub('apm:express:route:added', ({ topOfStackFunc, layer }) => {
+      if (Object.hasOwn(layer, kCodeOriginForSpansTagsSym)) return
+
+      layer[kCodeOriginForSpansTagsSym] = entryTags(topOfStackFunc, 1)
```

Comment:
After testing this on a real life Express app, I found that the top frame this `1` is skipping (`shimmer.wrapFunction`) isn't always present, in which case the code is skipping the stack frame we actually care about.

More research is needed to make this code more resilient. However, I'll mark this PR as "Draft" and table this work to focus on other more pressing issues.

Reply:
For some reason, the app used in tests no longer has the issue that required us to skip the top frame. I've not been able to reproduce the issue either. So without any more to go by, I've removed the "skip-top-frame-fix" and now just always pick the actual top frame as we would normally expect to do. I think this might be due to some of the recent changes to shimmer or how Express is being instrumented. If that's the case, we are safe to move on.